truffle/src/com.oracle.truffle.api/src/com/oracle/truffle/api/TruffleSafepoint.java

```diff
@@ -0,0 +1,399 @@
+     * Truffle to interrupt the lock temporarily to perform a thread local action. This method is
+     * safe to be used on compiled code paths.
+     * <p>
+     * Example usage:
+     * <p>
+     * Note there is a short-cut method to achieve a the same behavior as in this example
+     * {@link #setBlockedInterruptable(Node, ThreadInterruptable, Object)}.
+     *
+     * <pre>
+     * Lock lock = new ReentrantLock();
+     * TruffleSafepoint sp = TruffleSafepoint.getCurrent();
+     * Interrupter prev = sp.setBlocked(Interrupter.THREAD_INTERRUPT);
+     * try {
+     *     while (true) {
+     *         try {
+     *             lock.lockInterruptibly();
+     *             break;
+     *         } catch (InterruptedException e) {
+     *             TruffleSafepoint.poll(this);
+     *         }
+     *     }
+     * } finally {
+     *     sp.setBlocked(prev);
+     * }
+     * </pre>
+     *
+     * @see TruffleSafepoint
+     * @param interrupter the interrupter to interrupt the current thread
+     * @since 21.1
+     */
+    public abstract Interrupter setBlocked(Interrupter interrupter);
+
+    /**
+     * Short-cut method to allow setting the blocked status for methods that throw
+     * {@link InterruptedException} and support interrupting using {@link Thread#interrupt()}.
+     * <p>
+     *
+     * <pre>
+     * TruffleSafepoint.setBlockedInterruptable(location, Lock::lockInterruptibly, lock);
+     * </pre>
+     *
+     * This method is short-hand for:
+     *
+     * <pre>
+     * TruffleSafepoint sp = TruffleSafepoint.getCurrent();
+     * Interrupter prev = sp.setBlocked(Interrupter.THREAD_INTERRUPT);
+     * try {
+     *     while (true) {
+     *         try {
+     *             interruptable.run(lock);
+     *             break;
+     *         } catch (InterruptedException e) {
+     *             TruffleSafepoint.poll(this);
+     *         }
+     *     }
+     * } finally {
+     *     sp.setBlocked(prev);
+     * }
+     * </pre>
+     *
+     *
+     * @param location the location with which the safepoint should be polled.
+     * @param interruptable the thread interruptable method to use for locking the object
+     * @param object the instance to use the interruptable method with.
+     *
+     * @since 21.1
+     */
+    @TruffleBoundary
+    public static <T> void setBlockedInterruptable(Node location, ThreadInterruptable<T> interruptable, T object) {
+        TruffleSafepoint safepoint = TruffleSafepoint.getCurrent();
+        Interrupter prev = safepoint.setBlocked(Interrupter.THREAD_INTERRUPT);
+        try {
+            while (true) {
+                try {
+                    interruptable.apply(object);
+                    break;
+                } catch (InterruptedException e) {
+                    TruffleSafepoint.poll(location);
+                }
+            }
+        } finally {
+            safepoint.setBlocked(prev);
+        }
+    }
```

Comment:
I think this is where TruffleRuby does some things with signals to interrupt native blocking calls? @eregon

Reply:
The SIGVTALRM stuff is essentially just another `Interrupter` that we'd set around blocking NFI calls.
This method is basically equivalent to `ThreadManager#runUntilResult()`.